wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,309 @@
+
+    fn nonce_request_via_login(&self) -> Option<WpNetworkRequest> {
+        let mut headers = http::HeaderMap::new();
+        headers.insert(
+            http::header::CONTENT_TYPE,
+            HeaderValue::from_static(CONTENT_TYPE_FORM),
+        );
+
+        let body = serde_urlencoded::to_string([
+            ["log", self.credentials.username.as_str()],
+            ["pwd", self.credentials.password.as_str()],
+            ["rememberme", "true"],
+            [
+                "redirect_to",
+                self.api_base_url.derived_rest_nonce_url().as_str(),
+            ],
+        ])
+        .ok()?;
+
+        Some(WpNetworkRequest {
+            method: RequestMethod::POST,
+            url: self.api_base_url.derived_wp_login_url().into(),
+            header_map: WpNetworkHeaderMap::new(headers).into(),
+            body: Some(WpNetworkRequestBody::new(body.into_bytes()).into()),
+        })
+    }
+
+    async fn nonce_from_request(&self, request: WpNetworkRequest) -> Option<String> {
+        self.request_executor
+            .execute(request.into())
+            .await
+            .ok()
+            .and_then(|response| {
+                // A 200 OK response from the `rest_nonce_url` (a.k.a `wp-admin/admin-ajax.php`)
+                // should be the nonce value. However, just in case the site is configured to
+                // return a 200 OK response with other content (for example redirection to
+                // other webpage), here we check the body length here for a light validation of
+                // the nonce value.
+                if response.status_code == 200 {
+                    let body = response.body_as_string();
+                    if body.len() < 50 {
+                        return Some(body);
+                    }
+                }
+                None
+            })
+    }
+}
+
+#[async_trait::async_trait]
+impl Authenticator for CookieAuthenticator {
+    fn host(&self) -> &str {
+        self.api_base_url.host_str()
+    }
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool {
+        let is_same_host = Url::parse(request_url)
+            .ok()
+            .map(|f| f.host_str() == Some(self.host()));
+        if is_same_host != Some(true) {
+            return false;
+        }
+
+        if let Some(response_status_code) = response_status_code {
+            return response_status_code == 401;
+        }
+
+        true
+    }
+
+    async fn authentication_headers(&self) -> Option<HeaderMap> {
+        self.fetch_rest_nonce().await.map(|nonce| {
+            let mut headers = HeaderMap::new();
+            headers.insert(
+                "X-WP-Nonce",
+                nonce
+                    .try_into()
+                    .expect("This conversion should never fail since nonce is a short string"),
+            );
+            headers
+        })
+    }
+
+    async fn reset(&self) {
+        *self.nonce.lock().await = None;
```

Comment:
I wonder what will happen now with the `nonce` with the new concurrent handling when an authentication is going on and another thread resets the `nonce`. I _think_ it might end up removing the newly acquired `nonce`.

I could be wrong, but I feel like the removing the expired token should be part of the current `fetch_rest_nonce` implementation. With the current naming maybe this suggestion doesn't sound correct, but as for the implementation, I think it makes sense to keep the mutex modification to a single place. It sort of works like:

* If the cached token is expired, clear it
* If we have a valid token return it
* If we don't have a valid token, retrieve it, cache it and return it

Furthermore, I had some concerns about the `reset` function in general, because I felt it was a `CookieAuthenticator` specific thing. So, I was wondering if it would make sense to pass an argument to `authentication_headers` function indicating that it should clear its cached token. Maybe that way we don't need the `re_authenticate` either.

What do you think?

Reply:
> I feel like the removing the expired token should be part of the current fetch_rest_nonce implementation. 

The tricky thing is we don't know if a nonce is valid or not, until we make an HTTP request with it.

With the implementation in this PR, when there is an existing nonce in memory, the best scenario (which would be 90% of the time for the apps because nonce is valid for one day by default, which probably would outlive the app) is we only need to make one HTTP request, which is the REST API request, and the worst scenario (when nonce expires) is making three HTTP requests, which are one REST API request, one request to get the valid nonce, retry the REST API request.

> If the cached token is expired, clear it

We don't know when it expires. If we were to use an HTTP request to check if it's expired, we'd need to make an additional nonce-fetching HTTP request for every single REST API call, which I don't think is ideal. If we were to go with this solution( I don't think we should), we don't need to cache nonce at all.

Reply:
> We don't know when it expires

When I say, if the cached token is expired, I mean the case where we manually call the `reset` function.

Reply:
I have made another change to address your concerns: https://github.com/Automattic/wordpress-rs/pull/327/commits/86caf4247d2cbea88c9fc1ced4b3cae7c3073ba5

> I wonder what will happen now with the nonce with the new concurrent handling when an authentication is going on and another thread resets the nonce. I think it might end up removing the newly acquired nonce.

Great catch! The new implementation now checks the previously used nonce and only clears the cached nonce if they are the same. That means the second request that's waiting for the ongoing authentication request won't reset the nonce. I'm looking into adding an unit test to cover this scenario.

> I could be wrong, but I feel like the removing the expired token should be part of the current fetch_rest_nonce implementation.

This is also done as part of the above change. Also, `fetch_rest_nonce` is now renamed to `update_reset_nonce`. Maybe `update_rest_nonce_if_needed` would be a better name? But I'm not sure if that's too verbose...

> Furthermore, I had some concerns about the reset function in general, because I felt it was a CookieAuthenticator specific thing. 

Indeed. I have simplified the trait design but generalized `authentication_headers` arguments (changing URL and status code to request and response objects) to avoid coupling the trait API with one specific implementation.

Reply:
I have added some unit tests, but they are quite verbose and not very easy to read. Let me know if you have any thoughts in improving them.